jobmanager/WorkerThread.cpp

```diff
@@ -45,6 +95,68 @@ void WorkerThread::run()
     {
         logger::error("unknown exception", "WorkerThread");
     }
+    workerThreadHandler = nullptr;
+}
+
+// return true if any new jobs were processed
+bool WorkerThread::processJobs()
+{
+    if (_backgroundJobCount == _backgroundJobs.size())
+    {
+        _backgroundJobCount = processBackgroundJobs();
+        if (_backgroundJobCount == _backgroundJobs.size())
+        {
+            // do not take on more jobs as we cannot store them if in case they are be multi-step job
+            return false;
+        }
+    }
+
+    uint32_t processedJobs = 0;
+    for (processedJobs = 0; processedJobs < 10; ++processedJobs)
+    {
+        auto job = _jobManager.pop();
+        if (!job)
+        {
+            break;
+        }
+
+        bool runAgain = job->runStep();
+        if (!runAgain)
+        {
+            _jobManager.freeJob(job);
+        }
+        else
+        {
+            for (uint32_t i = 0; i < _backgroundJobs.size(); ++i)
+            {
+                if (!_backgroundJobs[i])
```

Comment:
Leaking jobs here. For example if _backgroundJobs only have space for 1 job and this cycle of 10 jobs need to add more than 1, the others will be leaked.